frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts

```diff
@@ -251,5 +284,249 @@ void main() {
+  // - the first and the last vertex isn't moved so that the plane keeps its original extent
+  // - the second last vertex is moved so that it's positioned at the beginning of the last
+  //   rendered bucket. If this isn't handled additionally, the second last vertex might be
+  //   still correct in most cases, but at certain positions, the second last vertex is
+  //   already in the second last bucket and would be moved to far to the left.
+  //   Example:
+  //   *------|------|------|----a-|---*
+  //   ^ first vertex
+  //          ^ second vertex
+  //                   ...             ^ last vertex
+  //                             ^ if a is the second last vertex, it needs to be moved to the right
+  //                               instead of the left.
+
+  // Calculate the index of the vertex (e.g., index.x=0 is the first horizontal vertex).
+  index = (position.xy / (PLANE_WIDTH / 2.) + 1.) / 2. * PLANE_SUBDIVISION;
+
+  uint activeMagIdx = uint(activeMagIndices[representativeLayerIdxForMag]);
+  vec2 d = transDim(vec3(32.) * getResolution(activeMagIdx)).xy;
+
+  vec3 datasetScaleUVW = transDim(datasetScale);
+  vec3 transWorldCoord = transDim(worldCoord.xyz);
+
+  if (index.x >= 1. && index.x <= PLANE_SUBDIVISION - 2.) {
+    transWorldCoord.x = floor(transWorldCoord.x / datasetScaleUVW.x / d.x) * d.x * datasetScaleUVW.x;
+    transWorldCoord.x = clamp(transWorldCoord.x, worldCoordTopLeft.x, worldCoordBottomRight.x);
+  } else if (index.x == PLANE_SUBDIVISION - 1.) {
+    // The second-last vertex should be clipped to the next-lower bucket boundary beginning from
+    // worldCoordBottomRight.
+    transWorldCoord.x = floor(worldCoordBottomRight.x / datasetScaleUVW.x / d.x) * d.x * datasetScaleUVW.x;
+  }
+
+  if (index.y >= 1. && index.y <= PLANE_SUBDIVISION - 1.) {
+    transWorldCoord.y = floor(transWorldCoord.y / datasetScaleUVW.y / d.y) * d.y * datasetScaleUVW.y;
+    transWorldCoord.y = clamp(transWorldCoord.y, worldCoordTopLeft.y, worldCoordBottomRight.y);
+  } else if (index.y == PLANE_SUBDIVISION - 1.) {
+    // The second-last vertex should be clipped to the next-lower bucket boundary beginning from
+    // worldCoordBottomRight.
+    transWorldCoord.y = floor(worldCoordBottomRight.y / datasetScaleUVW.y / d.y) * d.y * datasetScaleUVW.y;
+  }
+
+  worldCoord = vec4(transDim(transWorldCoord), 1.);
+  vec3 posRec = (modelInv * worldCoord).xyz;
+  gl_Position = projectionMatrix * modelViewMatrix * vec4(posRec, 1.0);
+
+  vec3 worldCoordUVW = getWorldCoordUVW();
+
+  // Offset the bucket calculation for the current vertex by a voxel to ensure
+  // that the provoking vertex (the one that is used by the flat varyings in
+  // the corresponding triangle) looks up the correct bucket. Otherwise,
+  // a rendering offset of 32 vx occurs.
+  worldCoordUVW.x -= 1.;
+  worldCoordUVW.y += 1.;
+
+  flatVertexPos = worldCoordUVW;
+  float NOT_YET_COMMITTED_VALUE = pow(2., 21.) - 1.;
+  const float bucketWidth = 32.;
```

Comment:
I'd suggest to use `BUCKET_WIDTH` via templating here as well. Same for the FALLBACK_COUNT in line 489.